src/librustc/infer/error_reporting/need_type_info.rs

```diff
@@ -136,6 +154,35 @@ impl<'a, 'tcx> InferCtxt<'a, 'tcx> {
             let expr = self.tcx.hir().expect_expr(body_id.hir_id);
             local_visitor.visit_expr(expr);
         }
+        let err_span = if let Some(pattern) = local_visitor.found_arg_pattern {
+            pattern.span
+        } else {
+            span
+        };
+
+        let ty_msg = match local_visitor.found_ty {
```

Comment:
The slight `match` duplication is to allow us to build the `err` early, and to remove some unnecessary branching.